hasher-matcher-actioner/hmalib/common/writebacker_models.py

```diff
@@ -0,0 +1,240 @@
+            )
+
+        print("subclasses", cls.__subclasses__())
+
+        sources_to_writebacker_cls = {
+            writebacker_cls().source: writebacker_cls
+            for writebacker_cls in cls.__subclasses__()
+        }
+
+        if source not in sources_to_writebacker_cls.keys():
+            return None
+        return sources_to_writebacker_cls[source]()
+
+    def writeback_is_enabled(self) -> bool:
+        """
+        Users can switch on/off writebacks either globally or for individual sources
+        """
+        raise NotImplementedError
+
+    @property
+    def writeback_type(self) -> WritebackLabel:
+        """
+        The writeback label for when this action should be performed (eg SawThisTooWritebackLabel())
+        """
+        raise NotImplementedError
+
+    def _writeback_impl(self, writeback_message: WritebackMessage) -> str:
+        raise NotImplementedError
+
+    def perform_writeback(self, writeback_message: WritebackMessage) -> str:
+        writeback_options = self.writeback_options()
+        writeback_to_perform = writeback_message.writeback_label.value
+        if writeback_to_perform not in writeback_options:
+            return (
+                "Could not find writebacker for source "
+                + self.source
+                + " that can perform writeback "
+                + writeback_to_perform
+            )
+
+        writebacker = writeback_options[writeback_to_perform]()
+        if writebacker.writeback_is_enabled:
+            return writebacker._writeback_impl(writeback_message)
+        return "Writeback {writebacker.__name__} not performed becuase it switched off"
+
+
+@dataclass
+class ThreatExchangeWritebacker(Writebacker):
+    """
+    Writebacker parent object for all writebacks to ThreatExchange
+    """
+
+    source = "te"
+
+    @staticmethod
+    @lru_cache(maxsize=None)
+    def writeback_options() -> t.Dict[str, t.Type["Writebacker"]]:
+        return {
+            "FalsePositive": ThreatExchangeFalsePositiveWritebacker,
+            "TruePositive": ThreatExchangeTruePositivePositiveWritebacker,
+            "Ingested": ThreatExchangeIngestedWritebacker,
+            "SawThisToo": ThreatExchangeSawThisTooWritebacker,
+        }
+
+    def writeback_is_enabled(self) -> bool:
+        """
+        TODO implement
+        Looks up from a config whether ThreatExchange writing back is enabled. Initially this will be a global
+        config, and this method will return True if writing back is enabled, False otherwise. At some point the
+        config for writing back to ThreatExchange may be on a per collaboration basis. In that case, the config
+        will be referenced for each collaboration involved (implied by the match message). If writing back
+        is enabled for a given collaboration, a label will be added to the match message
+        (e.g. "ThreatExchangeWritingBackEnabled:<collaboration-id>").
+        """
+        return True
```

Comment:
I'm remembering that the team already decided that the granularity of enabling reactions to TE is at the collaboration level, hence Mengyang's work to add the toggle on a per collaboration basis. Please update the TODO comment to reflect this. When we implement, we just need to get all the configs and use that. But this will make things more complicated... You'll need to pass in the message and output a dictionary of "data set id" to bool to use later in what gets written back. I think you should do some deeper analysis of how your dynamic code works in this case. Will the necessary output of this method flow through?

Reply:
Updated the comment

I see your point about the refactor however I dont think it will be a heavy lift and I'd like to get this PR our soon